_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+
+<!-- excerpt start -->
+
+In this article, I discuss some of the problems I’ve encountered in both producing and analyzing device logs over the years, on single digit number of devices up to hundred of thousands, and concepts I’ve applied to manage that complexity. My background is in Android development, so some of the examples may skew towards much more powerful hardware than simple microcontrollers.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+---
+
+## Log Collection
+
+As is common in embedded development, there are many interesting decisions to be made long before any code is actually written, or any logs actually logged. Before we can use the logs, we first need to design how the logs are stored on the device, and then how to get access to them.
+
+### Log Cache Size
+
+One of the fundamental constraints to work around with device logging is how much disk space to allocate just for logging. Unlike on servers where the log cache is basically infinite (and you can often pay more for the storage), most logging systems on hardware will reserve a fixed amount of disk space or memory in the form of a rotating buffer of logs. When you log past the limit, the logs in the oldest buffer gets written over.
+
+The amount of disk space chosen has a large influence on logging &mdash; it affects how verbose your logs can be, how long your logs are be retained, what information you'll want prioritize logging, how you'll store the logs, and more. You’ll want to choose the maximum amount of space you can afford, where the logs you get back contain enough useful information at the timeframe of the incident. But you'll also want to change how you log to accomodate for the amount of space, such as using an abbreviated short form instead of full words.
+
+I've been lucky in that embedded Android devices are generally not very disk space constrained, so I’ve had the benefit of having 10's of MBs of logs to store (and comb) through. But even so, that would only constitute maybe 2-3 hours of logs, depending on how busy the device was, and there was a constant trade-off with how much information was being logged.
+
+### On-location v. Remote Collection
+
+One of my favorite often overlooked questions when you develop a new embedded device product is, "how do I get the logs off of my production device and onto my computer"? During development, this could be as easy as connecting a cable directly to the device and downloading the logs files off the filesystem directly. But when the device is in the field, it’s not so simple.
```

Comment:
We jump from Log Collection Cache Size directly to Log Collection. Is it worth adding a section talking about the pros and cons of in-memory vs writing to disk? Android defaults to in-memory. MCU (from my very limited experience) will often times *just* stream them over serial. The main things to balance that I can think of currently are memory constraints vs disk wear.

Reply:
I'm not sure in memory vs streaming makes sense in the context of this article, because if it's not persisted to the filesystem, then there's no real way of remote debugging it? I also kind of prefaced that the article has an Android slant.

But I do like that it adds some more weight to the kind of bare section.

Reply:
lgtm